codex-rs/core/src/admin_controls.rs

```diff
@@ -0,0 +1,391 @@
+    let line = serde_json::to_string(record).map_err(io::Error::other)?;
+    file.write_all(line.as_bytes())?;
+    file.write_all(b"\n")?;
+    Ok(())
+}
+
+fn send_record_to_endpoint(endpoint: &str, record: AdminAuditRecord) {
+    match Handle::try_current() {
+        Ok(handle) => {
+            let client = reqwest::Client::new();
+            let endpoint = endpoint.to_string();
+            handle.spawn(async move {
+                if let Err(err) = client.post(endpoint).json(&record).send().await {
+                    warn!(?err, "failed to post admin audit event");
+                }
+            });
+        }
+        Err(_) => {
+            warn!("admin audit HTTP logging requested but no async runtime is available");
+        }
+    }
+}
+
+fn expand_path(raw: &str) -> io::Result<PathBuf> {
+    if raw == "~" {
+        return dirs::home_dir().ok_or_else(|| {
+            io::Error::new(
+                io::ErrorKind::NotFound,
+                "could not resolve home directory for admin audit log file",
+            )
+        });
+    }
+
+    if let Some(rest) = raw.strip_prefix("~/") {
+        let mut home = dirs::home_dir().ok_or_else(|| {
+            io::Error::new(
+                io::ErrorKind::NotFound,
+                "could not resolve home directory for admin audit log file",
+            )
+        })?;
+        if !rest.is_empty() {
+            home.push(rest);
+        }
+        return Ok(home);
+    }
+
+    Ok(PathBuf::from(raw))
+}
+
+fn current_username() -> String {
+    env_var("USER")
+        .or_else(|| env_var("USERNAME"))
+        .unwrap_or_else(|| "unknown".to_string())
+}
+
+fn current_hostname() -> String {
+    gethostname()
+        .into_string()
+        .ok()
+        .filter(|value| !value.is_empty())
+        .or_else(|| env_var("HOSTNAME"))
+        .or_else(|| env_var("COMPUTERNAME"))
+        .unwrap_or_else(|| "unknown".to_string())
+}
+
+fn env_var(key: &str) -> Option<String> {
+    std::env::var(key).ok().filter(|value| !value.is_empty())
+}
+
+fn sandbox_label(policy: &SandboxPolicy) -> &'static str {
+    match policy {
+        SandboxPolicy::DangerFullAccess => "danger-full-access",
+        SandboxPolicy::ReadOnly => "read-only",
+        SandboxPolicy::WorkspaceWrite { .. } => "workspace-write",
+    }
+}
+
+fn sandbox_type_label(sandbox: SandboxType) -> &'static str {
+    match sandbox {
+        SandboxType::None => "none",
+        SandboxType::MacosSeatbelt => "macos-seatbelt",
+        SandboxType::LinuxSeccomp => "linux-seccomp",
+    }
+}
+
+pub fn build_danger_audit_payload(
+    pending: &DangerPending,
+    action: DangerAuditAction,
+    justification: Option<String>,
+) -> AdminAuditPayload {
+    AdminAuditPayload::Danger(DangerAuditDetails {
+        action,
+        justification,
+        requested_by: pending.source,
+        sandbox: sandbox_label(&pending.requested_sandbox).to_string(),
```

Comment:
I would keep the type as `SandboxType` rather than convert to `String`. Only do that when you actually have to serialize, and even then, serde can help you.